framework/src/base/MooseApp.C

```diff
@@ -1163,6 +1178,162 @@ MooseApp::disableCheckUnusedFlag()
   _enable_unused_check = OFF;
 }
 
+FEProblemBase &
+MooseApp::feProblem() const
+{
+  return _executor.get() ? _executor->feProblem() : _executioner->feProblem();
+}
+
+void
+MooseApp::addExecutor(const std::string & type,
+                      const std::string & name,
+                      const InputParameters & params)
+{
+  std::shared_ptr<Executor> executor = _factory.create<Executor>(type, name, params);
+
+  if (_executors.count(executor->name()) > 0)
+    mooseError("an executor with name '", executor->name(), "' already exists");
+  _executors[executor->name()] = executor;
+}
+
+void
+MooseApp::addExecutorParams(const std::string & type,
+                            const std::string & name,
+                            const InputParameters & params)
+{
+  _executor_params[name] = std::make_pair(type, libmesh_make_unique<InputParameters>(params));
+}
+
+void
+MooseApp::recursivelyCreateExecutors(const std::string & current_executor_name,
```

Comment:
I would simply have a parent executor to create its own children. The job of MooseApp is add the highest level executor. 

This idea will allow a customized executor to define its fashion to create its children.   Another advantage of this idea is that we then allow a delayed creation of children, which might be useful for some applications.

Reply:
This is just for creating the objects.  They do "fetch" their own children within themselves.  They plumb themselves together - this is mostly just for creating them in the right order so we can bind them together in their constructors - we need to handle deps so we know which ones to create first, etc.

Reply:
One Executor can be the child of multiple others.

Reply:
>This is just for creating the objects.

I understood. How about I do not want to create my children executors when the outtest extupor is created?  

I could decide if I want to create an inner solver or not based on the result from the previous solver.  For example, if the init solver alway fails, why I need to waste memory to create the following-up solvers?